pkg/deployable/protobuf_translator.go

```diff
@@ -0,0 +1,225 @@
+// Copyright 2025 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deployable
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/okteto/okteto/cmd/utils"
+	"github.com/okteto/okteto/pkg/divert"
+	"github.com/okteto/okteto/pkg/k8s/labels"
+	oktetoLog "github.com/okteto/okteto/pkg/log"
+	"github.com/okteto/okteto/pkg/model"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/protobuf"
+	"k8s.io/kubectl/pkg/scheme"
+)
+
+type protobufTranslator struct {
+	serializer   *protobuf.Serializer
+	divertDriver divert.Driver
+	name         string
+}
+
+func newProtobufTranslator(name string, divertDriver divert.Driver) *protobufTranslator {
+	protobufSerializer := protobuf.NewSerializer(scheme.Scheme, scheme.Scheme)
+	return &protobufTranslator{
+		name:         name,
+		serializer:   protobufSerializer,
+		divertDriver: divertDriver,
+	}
+}
+
+func (p *protobufTranslator) Translate(b []byte) ([]byte, error) {
+	// Passing nil for defaultGVK and into, so the serializer infers the GVK from the data and creates a new object.
+	// This is necessary because the object is not known at compile time.
+	obj, _, err := p.serializer.Decode(b, nil, nil)
+	if err != nil {
+		oktetoLog.Infof("error unmarshalling resource body on proxy: %s", err.Error())
+		return nil, fmt.Errorf("could not unmarshal resource body: %w", err)
+	}
+
+	if err := p.translateMetadata(obj); err != nil {
+		return nil, err
+	}
+
+	var buf bytes.Buffer
+	if err := p.serializer.Encode(obj, &buf); err != nil {
+		return nil, fmt.Errorf("could not encode resource: %w", err)
+	}
+
+	switch obj.GetObjectKind().GroupVersionKind().Kind {
```

Comment:
What about VirtualService translation?

Reply:
VirtualService is not implementing `runtime.Object` which prevents me to get the spec from the protobuf object.

Reply:
ok, but it is something we should keep an eye on. If not, divert with virtual services might stop working. But for now, it seems that client won't change the content type used